### ISSUE TITLE
Bug with the envelope generator

### ISSUE THREAD

**Issue by ValleyBell** [digital-sound-antiques/emu2149#2]
There seems to be a bug with the envelope generator that causes the channel to go "full volume" when it is not supposed to be.

It occours in the Level 1 music of Jim Power in Mutant Planet (Atari ST), on the first channel: [VGM + MP3](https://github.com/digital-sound-antiques/emu2149/files/7233800/Jim_Power_.Atari_ST._-_Level_1.zip)
The MP3 has EMU2149 on the left speaker and the MAME AY8910 sound core on the right speaker.
The bug can be heard at 00:16.26 in the VGM / 00:02.68 in the MP3.

The command sequence that appears to trigger the bug, is:
```
Channel state before:
0x00002CB1: A0 07 0C    AY8910:		Enable: Channel AB-, Noise -BC, Port II
0x00002CCE: A0 0B 00    AY8910:		Envelope Freq. Fine: 0x00
0x00002CD1: A0 0D 08    AY8910:		Envelope Shape: 0x08
0x00002CD5: A0 00 DD    AY8910:		Chn A Freq. Fine: 0xDD
0x00002CD9: A0 01 01    AY8910:		Chn A Freq. Coarse: 0x1
0x00002CDD: A0 08 1E    AY8910:		Chn A Volume 93%, Envelope Mode: 8

commands that seem to trigger the bug:
0x00002CF8: 71          Wait:	 2 sample(s) (   0.05 ms)	(total	717236 (00:16.26))
0x00002CF9: A0 0B FF    AY8910:		Envelope Freq. Fine: 0xFF
0x00002CFC: A0 0C FF    AY8910:		Envelope Freq. Coarse: 0xFF
0x00002CFF: A0 0D 08    AY8910:		Envelope Shape: 0x08
```

I also tried to do a minimal example that shows the same bug: [SSG-test.vgm.gz](https://github.com/digital-sound-antiques/emu2149/files/7233781/SSG-test.vgm.gz)
This is based on how the bug occours in the game "Dragon Half" (NEC PC-9801).

Thanks go to GTheGuardian from vgmrips for noticing the issue.

**Comment by ghost** [comment]
Might want to add in that similar issue is also present in the Atari ST version of After Burner and is an even more egregious example of this bug. ([Example VGM](https://www.dropbox.com/s/l4tltkcg9f0y3si/02%20Red%20Out.vgm?dl=1) / [MP3; Volume Warning on EMU2149 recording](https://www.dropbox.com/s/pnszrm4ywouxie6/RedOut.zip?dl=1) / [HW Reference](https://www.youtube.com/watch?v=Q0Av950xE8w&t=401s))

**Comment by okaxaki** [comment]
Thank you for detailed bug reports. I will investigate the problem and will fix it.

**Comment by okaxaki** [comment]
Hi, Folks,

I have fixed the bug and released [1.30](https://github.com/digital-sound-antiques/emu2149/releases/tag/1.30).
Thank you very much again for your reports!

### Some details
There were two problems behind this bug.

The first problem was that the envelope frequency 0 was handled incorrectly. The frequency 0 should have been the same as 1 but it haven't. I fixed this with the patch at [c48ec00](https://github.com/digital-sound-antiques/emu2149/commit/c48ec00202f90f58dcf5647fb2a0a9b5ecb1990c) and the patch fixes the incorrect noise sound of SSG-test.vgm and After Burner music.

The second problem was that the envelope counter was not latched immediately after R#11 and R#12 access. The patch for this problem is [31ee5a4](https://github.com/digital-sound-antiques/emu2149/commit/31ee5a4042ebc45b64d291bac0a74ebc0e725a15). This fixes "full volume" tones in Jim Power in Mutant Planet.

**Comment by ValleyBell** [comment]
I updated emu2149 in libvgm and I can confirm that the bug is fixed now.

Thanks!
